File: clojure/clock/_iterations/2/src/clock.clj
(ns clock)

(def minutes-per-hour 60)
(def minutes-per-day (* 24 minutes-per-hour))

(defn clock->string [c]
  (format "%02d:%02d" (quot c minutes-per-hour) (rem c minutes-per-hour)))

(defn clock [hours minutes]
  (mod (+ (* hours minutes-per-hour) minutes) minutes-per-day))

(defn add-time [c minutes-diff]
  (clock 0 (+ c minutes-diff)))

File: kotlin/clock/_iterations/2/src/main/kotlin/Clock.kt
data class Clock(private var _hours: Int, private var _minutes: Int) {
    init {
        adjust()
    }

    val hours: Int get() = _hours
    val minutes: Int get() = _minutes

    fun add(minutesDelta: Int) {
        _minutes += minutesDelta
        adjust()
    }

    fun subtract(minutesDelta: Int) {
        add(-minutesDelta)
    }

    override fun toString(): String {
        return "${hours.toLeadingZeroString()}:${minutes.toLeadingZeroString()}"
    }

    private fun adjust() {
        while (_minutes < 0) {
            --_hours
            _minutes += 60
        }
        while (_minutes >= 60) {
            ++_hours
            _minutes -= 60
        }
        while (_hours < 0) {
            _hours += 24
        }
        while (_hours >= 24) {
            _hours -= 24
        }
    }

    private fun Int.toLeadingZeroString(): String = if (this < 10) "0$this" else toString()
}
